### PR TITLE
F# RFC FS-1015 - Support for "fixed"

### PR DESCRIPTION
**Pull request by dsyme**
This is the PR for [RFC  FS-1015 - Support for 'fixed'](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1015-support-for-fixed.md)

* [x] Complete implementation 
* [x] Add tests
* [x] Fix baselines

Some tests are present but more are needed.

### CODE REVIEW

File: build.cmd [hunk 1]
```diff
@@ -153,18 +153,42 @@ if /i '%ARG%' == 'ci_part1' (
...
+    set TEST_FSHARPQA_SUITE=0
     set TEST_FSHARP_SUITE=1
+    set TEST_VS=0
+    set TEST_TAGS=
+)
+
+if /i '%ARG%' == 'ci_part3' (
+    set BUILD_PROTO=1
+    set SKIP_EXPENSIVE_TESTS=1
+    set BUILD_CORECLR=1
```
Comment (otawfik-ms) [inline build.cmd]: Why are we building CORECLR without testing it? shouldn't it be already covered in ci_part2?

**Comment by dsyme** [comment]
This is now ready for review and commit

File: src/fsharp/lexhelp.fs [hunk 2]
```diff
@@ -284,7 +285,7 @@ module Keywords =
         [ "atomic"; "break"; 
           "checked"; "component"; "constraint"; "constructor"; "continue"; 
           "eager"; 
-          "fixed"; "fori"; "functor";
```
Comment (vladima) [inline src/fsharp/lexhelp.fs]: yay! :)

**Comment by dsyme** [comment]
OK, this is now green - please review!

**Comment by codymack** [comment]
It looks like, for `str` a `string`, that both

`use pStr = fixed str`
`use (pStr: nativeptr<char>) = fixed str`

compile to the `native int` type in IL, whereas the C# `fixed (char* pStr = str)` compiles to the `char*` type in IL.

Not being an expert in how the CLR works, is this a cause for concern?

File: src/absil/ilwrite.fs [hunk 3]
```diff
@@ -1494,7 +1500,10 @@ let GetCallsigAsStandAloneSigIdx cenv env info =
 let EmitLocalSig cenv env (bb: ByteBuffer) (locals: ILLocals) = 
     bb.EmitByte e_IMAGE_CEE_CS_CALLCONV_LOCAL_SIG
     bb.EmitZ32 locals.Length
-    locals |> ILList.iter (fun l -> EmitType cenv env bb l.Type) 
+    locals |> ILList.iter (fun l -> 
+        if l.IsPinned then 
+            bb.EmitByte et_PINNED
+        EmitType cenv env bb l.Type)
```
Comment (mexx) [inline src/absil/ilwrite.fs]: Can we please remove the duplication for this logic?
How about a new function `EmitTypeOfLocal`?

File: src/fsharp/tast.fs [hunk 4]
```diff
@@ -110,133 +110,140 @@ type ValFlags(flags:int64) =
...
+
+
     /// Get the flags as included in the F# binary metadata
     member x.PickledBits = 
         // Clear the RecursiveValInfo, only used during inference and irrelevant across assembly boundaries
         // Clear the IsCompiledAsStaticPropertyWithoutField, only used to determine whether to use a true field for a value, and to eliminate the optimization info for observable bindings
         // Clear the HasBeenReferenced, only used to report "unreferenced variable" warnings and to help collect 'it' values in FSI.EXE
         // Clear the IsGeneratedEventVal, since there's no use in propagating specialname information for generated add/remove event vals
-                                                      (flags       &&&    ~~~0b011001100000000000L) 
+                                                      (flags       &&&    ~~~0b1011001100000000000L)
```
Comment (mexx) [inline src/fsharp/tast.fs]: If I understand correctly we clear the bit for IsFixed, right?
Can we add a comment about the reasons as well?

File: src/fsharp/Optimizer.fs [hunk 5]
```diff
@@ -431,7 +431,7 @@ let BindExternalLocalVal cenv (v:Val) vval env =
     CheckInlineValueIsComplete v vval;
 #endif
 
-    if verboseOptimizationInfo then dprintn ("*** Binding "^v.LogicalName); 
+    if verboseOptimizationInfo then dprintn ("*** Binding "+v.LogicalName);
```
Comment (mexx) [inline src/fsharp/Optimizer.fs]: how is this change of operator is related to the feature?

File: src/fsharp/tast.fs (hunk 4 above)
Comment (dsyme) [inline src/fsharp/tast.fs]: Good catch, Ii had the bit the wrong way around, it should be kept

File: src/absil/ilwrite.fs (hunk 3 above)
Comment (dsyme) [inline src/absil/ilwrite.fs]: thanks, will do

File: src/fsharp/Optimizer.fs (hunk 5 above)
Comment (dsyme) [inline src/fsharp/Optimizer.fs]: it's not, just cleanup. We're gradually being a bit more relaxed about that.

**Comment by dsyme** [comment]
@codymack As far as I understand things that's not a cause for concern.  F# has always used ``nativeint`` for pointers, at least within method body code - but operationally there is no difference between the two.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
@dsyme  can you merge this

**Pull request merged**